[Minor] Custom function input parameter supports CellValue

xai-msft opened the pull request:
This is to support using [cell values](https://learn.microsoft.com/en-us/office/dev/add-ins/excel/excel-data-types-concepts) as custom function parameters.
We generate the corresponding cell value type in functions.json, and we throw errors for unsupported types.
Cell value types in return type will get converted to "any".

1. **Do these changes impact *command syntax* of any of the packages?** (e.g., add/remove command, add/remove a command parameter, or update required parameters)
    No.


2. **Do these changes impact *documentation*?** (e.g., a tutorial on https://learn.microsoft.com/office/dev/add-ins/overview/office-add-ins)
    We are not announcing cell value support here. No for now.


**Validation/testing performed**:

Added tests and ran tests.

donlvMSFT commented on packages/custom-functions-metadata/test/cases/input-parameter-types-cellvalue/functions.js:
```diff
@@ -0,0 +1,148 @@
+
+/**
+ * Test Excel.BooleanCellValue repeating matrix parameter
+ * @param {Excel.BooleanCellValue[][][]} x
```
Just curious when could CF have this kind of array? As data in the worksheet is 2d, we considering user just give input manually?

xai-msft replied:
I didn't know this before either, but apparently [][][] can be used for repeating range parameters. See https://learn.microsoft.com/en-us/office/dev/add-ins/excel/custom-functions-parameter-options?tabs=javascript .

akrantz replied:
https://learn.microsoft.com/en-us/office/dev/add-ins/excel/custom-functions-parameter-options?tabs=javascript#declaring-repeating-parameters

Does custom-functions-metadata support repeating scalar or matrix parameters?  The help mentions hand-writing the functions.json metadata to have `repeating: true`. It would be good to support it so it can be determined from the dimensionality of the array type, e.g. `number[][][]`. This would not be for this PR but something else to add to the backlog.

akrantz replied:
It is supported.  

https://github.com/OfficeDev/Office-Addin-Scripts/blob/fb1b94d0e3654680aa67088d098f4719e94484cd/packages/custom-functions-metadata/test/cases/repeating-param-3d/functions.ts#L5

xai-msft commented on packages/custom-functions-metadata/test/cases/error-unsupported-input-parameter/functions.ts:
```diff
@@ -0,0 +1,10 @@
+// Licensed under the MIT license.
+
+/**
+ * Test unsupported parameter
```
I wanted to give a different error message for unsupported cell values to reduce confusion / to convey the message that "we support cell values, just not this one". However, I can also just not distinguish between these two cases. What do you think?

akrantz commented:
@xai-msft Has the spec been updated? I would like to better understand the set of changes being made here before reviewing them.

akrantz commented on packages/custom-functions-metadata/src/parseTree.ts:
```diff
@@ -619,6 +640,14 @@ function getResults(
     type: resultType,
   };
 
+  // We only support cell value types as input and not return type for now.
```
Not sure why you need to give an error if a cell value type is put for a return value. It will still just generate "any" for it since the return type info in the metadata. 

Later it would be good to tighten this, e.g. the linked entity provider function should return `Promise<CellValue>` where CellValue is any cell value type (including error types).

xai-msft replied:
Please see my other comment, thanks.

akrantz commented on packages/custom-functions-metadata/src/parseTree.ts:
```diff
@@ -138,6 +138,27 @@ const TYPE_CUSTOM_FUNCTION_CANCELABLE = {
+  "Excel.BooleanCellValue": "booleancellvalue",
+  "Excel.DoubleCellValue": "doublecellvalue",
+  "Excel.EntityCellValue": "entitycellvalue",
+  "Excel.ErrorCellValue": "errorcellvalue",
```
Aren't there other error cell value types? They should all be supported.

It would be nice to have type info which could tell that the type is derived from Excel.CellValue, but currently there is no type info for these types with custom-functions-metadate. If the types are pulled out into an excel.cellvalue.d.ts for the type info, then that could be used by the code here and perhaps have better knowledge about the type hierarchy, or be written where new types could be added and the checks in here can just tell whether the class derives from Excel.CellValue. But this approach based on the list of type names is OK for now.

xai-msft replied:
We don't support error subtypes now.

akrantz commented on packages/custom-functions-metadata/test/cases/input-parameter-types-cellvalue/expected.json:
```diff
@@ -0,0 +1,205 @@
+    "allowCustomDataForDataTypeAny": true,
+    "functions": [
+        {
+            "description": "Test Excel.CellValue parameter",
```
There is currently only one successful case which is for Excel.CellValue. Really, there should be a case for each type.

xai-msft replied:
Sorry, could you clarify your comment? I have all the supported subtypes of CellValue here. What other type do you mean?

akrantz replied:
Each of the *CellValue types.

xai-msft replied:
I hope I'm not missing something but I have all the *CellValue types here? If one of them fails the whole test will fail.

akrantz commented on packages/custom-functions-metadata/test/cases/error-unsupported-return-type-cellvalue/functions.ts:
```diff
@@ -0,0 +1,10 @@
+ * Cell Value types are not supported as return types
+ * @customfunction
+ */
+function testUnsupportedCellValue(x): Excel.BooleanCellValue {
```
As per my other comment, I think this can/should be relaxed. The function can return a cell value or error cell value type. For the linked entity provider function, it MUST return Promise<Excel.CellValue> since it needs to return the linked entity.

xai-msft replied:
See my other comment. Dev should continue to use Promise<any> at this stage until we actually support return type.

xai-msft replied:
Metadata can contain boolean/string/number as return types. Excel used to do type check on these but it no longer does for some reason.

That said, we had some discussions, and we agree that converting CellValues as return types to "any" is better. I'll update the PR.

akrantz reviewed (changes requested):
This is a good start. Please see my comments. Please also update the spec so it supports the changes.

xai-msft commented:
> @xai-msft Has the spec been updated? I would like to better understand the set of changes being made here before reviewing them.

Thank you for your comments!
I'd like to sum up a few things first before addressing the comments individually:

1. We do not have return value CellValue support on Excel parsing side, which means if we include any new types in return parameter, the whole add-in registration will fail. We really cannot add them in metadata at this stage.
To support return value means to implement both registration and validation logic, which is a bigger work item on its own. We will do input / return in iterations as we agreed with in the spec discussion.

2. One thing this PR deviates from the spec is that while in the spec I decided to convert unsupported CellValue types (Reference etc.) to "any", I decided in this PR that silently failing is not a good idea. The rest is the same as in the spec. I'll update it to reflect this design change. Thanks for the reminder!

4. Error for unsupported return type: it wasn't specified in the spec since the spec was about input type. For the same reason as above, I think silently failing by converting CellValue types into "any" is not ideal, since it might give dev the wrong impression that it persists and we do validation on Excel like with the rest of the supported types, which we don't. I have to filter it out in the generator since both input and return go through this same utility function. I can update the spec to include the information that metadata generator throws error on CellValue as return type too.

I hope this answers your questions and please let me know if there's further concern!

xai-msft commented:
@akrantz Hi, I've updated the spec. Feel free to let me know what you think of my responses to your comments when you have time. Thank you!

akrantz commented:
@xai-msft I ran out of time looking at this right now. Might be good to meet up and discuss rather than going back and forth in comments. Will look more later. Let me know when is a good time to talk.

akrantz commented:
@xai-msft I don't see the specification.md file added to the PR. Can you check this please?

xai-msft commented:
> @xai-msft I don't see the specification.md file added to the PR. Can you check this please?

Sorry I misunderstood. I thought you were talking about the spec doc on internal SharePoint that you previously reviewed.

We are not announcing this feature yet since Excel code is still rolling out. I'd like to update the spec here (together with official documentation) after the rollout completes. And I'll create a work item to track the spec update here. Would you be ok with that?

xai-msft commented:
gently pinging @akrantz

To sum up:
1. CellValue type as return type has been changed to converting to "any" as requested
2. We want to postpone specification.md change till officially announcing the feature, since Excel code is still rolling out. 

Please let me know if there's other concerns. Thank you!

akrantz commented on packages/custom-functions-metadata-plugin/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "custom-functions-metadata-plugin",
-  "version": "1.6.2",
```
@millerds This version update should be done by Lerna, right? I don't think the version should be manually updated, but cannot quite remember.

millerds replied:
Correct . . . do not update the version, that will be done as part of the publish process.

xai-msft replied:
Reverted.

akrantz commented on packages/custom-functions-metadata/src/parseTree.ts:
```diff
@@ -1029,50 +1055,84 @@ function hasInvocationParameter(
-      }
-      arrayType = getArrayDimensionalityAndType(t);
-      kind = arrayType.type;
+  // Only get type for typescript files. js files will return "any" for all types
```
When using JavaScript, the user can annotate the type in the @param field, so it should be able to have a type when annotated for JavaScript, and we would expect this to be done for JavaScript in order to know the type.

Perhaps you want to add an error unit test case when a param in JavaScript doesn't specify a type, or perhaps this is the same any actually specifying `any` or `unknown` as the type in TypeScript. Might be useful to have test cases to clarify this.

In terms of the code here, it seems that not having a node implies that the type to return is any, so the code looks fine.

xai-msft replied:
Resolving as no issue present.

akrantz commented on packages/custom-functions-metadata/src/parseTree.ts:
```diff
@@ -1029,50 +1055,84 @@ function hasInvocationParameter(
+      // @ts-ignore
+      let cellValue = CELLVALUETYPE_MAPPINGS[nodeTypeName]
+      if (cellValue === "unsupported") {
+        extra.errors.push(logError("Custom function does not support cell value type: " + typeReferenceNode.typeName.getText(), typePosition));
```
Suggestion, use back-tick string instead of concatenating strings with +.

xai-msft replied:
Changed.

akrantz commented on packages/custom-functions-metadata/src/parseTree.ts:
```diff
@@ -1029,50 +1055,84 @@ function hasInvocationParameter(
+      return cellValue;
     }
+
+    extra.errors.push(logError("Invalid type: " + typeReferenceNode.typeName.getText(), typePosition));
```
same suggestion here to use back-tick string.

akrantz commented on packages/custom-functions-metadata/src/parseTree.ts:
```diff
@@ -1029,50 +1055,84 @@ function hasInvocationParameter(
+  // @ts-ignore
+  type = TYPE_MAPPINGS[node.kind];
+  if (!type) {
+    extra.errors.push(logError("Type doesn't match mappings", typePosition));
```
This error message is describing the implementation (using the type mappings array). Please compose an error message meaningful to the user. For example, in ScriptLab, if this error was encountered, what is really the problem to tell the user. Is it that they used a type that isn't supported?

xai-msft replied:
Hi, I moved the code around a bit for readability, but this error message was created by someone else. I'm not sure if I understand what this is for, or if the code ever hits, and I prefer not changing it for safety.

akrantz commented on packages/custom-functions-metadata/src/parseTree.ts:
```diff
@@ -1029,50 +1055,84 @@ function hasInvocationParameter(
+  if (ts.isTypeReferenceNode(node)) {
+    return (node as ts.TypeReferenceNode).typeName.getText() === "Array";
+  }
+  return false
```
semicolon is missing at the end of the line.

akrantz replied:
Make sure to run prettier if you're not having prettier applied in VS Code.

xai-msft replied:
Thanks. I ran prettier this iteration.

akrantz commented on packages/custom-functions-metadata/src/parseTree.ts:
```diff
@@ -1029,50 +1055,84 @@ function hasInvocationParameter(
   let array: IArrayType = {
     dimensionality: 0,
-    type: ts.SyntaxKind.AnyKeyword,
+    node: node,
```
You can use put; `node,` since the names are the same.

akrantz replied:
There are other places like this which can be updated in the same way.

xai-msft replied:
Thanks. I didn't know this syntax!

akrantz commented on packages/custom-functions-metadata/src/parseTree.ts:
```diff
@@ -1029,50 +1055,84 @@ function hasInvocationParameter(
+  if (typeNodeIsArray(node)) {
+    let arrayType: IArrayType = {
+      dimensionality: 0,
+      node: node,
```
You can just put `node,` since the names are the same.

akrantz commented on packages/custom-functions-metadata/src/parseTree.ts:
```diff
@@ -1083,49 +1143,47 @@ function getArrayDimensionalityAndType(node: ts.TypeNode): IArrayType {
-  array.type = nodeCheck.kind;
+  array.node = nodeCheck
   while (ts.isArrayTypeNode(nodeCheck)) {
     array.dimensionality++;
```
nit: I think it's preferrable to put ++ at the beginning.

akrantz commented on packages/custom-functions-metadata/test/cases/error-unsupported-input-parameter/expected.ts.errors.txt:
```diff
@@ -0,0 +1 @@
+Invalid type: Excel.Worksheet (8,40)
```
Rather than the message being `Invalid type`, it might be better to say that the type is not allowed for a custom function parameter.

akrantz reviewed (approved):
I've made some suggestions which I think would be good to make. I will mark as Approved so I don't block completion.

millerds commented on packages/custom-functions-metadata/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "custom-functions-metadata",
-  "version": "1.6.2",
```
do not update the version, that will be done as part of the publish process.

xai-msft replied:
reverted

akrantz reviewed (approved):
Thanks for making the updates!

xai-msft commented:
/azurepipelines run

xai-msft commented:
> Thanks for making the updates!

Thanks! Could you help me run `/azurepipelines run` ?

millerds commented:
/azurepipelines run
